src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+             {
+                 securityChannel = new SecurityReplySessionChannel(listener, (IReplySessionChannel)innerChannel, securityProtocol, listener.settingsLifetimeManager);
+             }
+             else
+             {
+                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.GetString(SR.UnsupportedChannelInterfaceType, typeof(TChannel))));
+             }*/
+
+             return (IChannel)securityChannel;
+        }
+
+    }
+
+   abstract class ServerSecurityChannel<UChannel>  where UChannel : class, IChannel
+    {
+        static MessageFault secureConversationCloseNotSupportedFault;
+        string secureConversationCloseAction;
+        SecurityListenerSettingsLifetimeManager settingsLifetimeManager;
+        bool hasSecurityStateReference;
+
+        internal SecurityProtocol SecurityProtocol { get;  set; }
+        public IReplyChannel OuterChannel { get; private set; }
+
+        protected ServerSecurityChannel( IReplyChannel outerChannel)
+            //: base(channelManager, innerChannel, securityProtocol) // not sure use of channel manager at this time
+        {
+            this.OuterChannel = outerChannel;
+        }
+
+
+        static MessageFault GetSecureConversationCloseNotSupportedFault()
+        {
+            if (secureConversationCloseNotSupportedFault == null)
+            {
+                FaultCode faultCode = FaultCode.CreateSenderFaultCode(DotNetSecurityStrings.SecureConversationCancelNotAllowedFault, DotNetSecurityStrings.Namespace);
+                FaultReason faultReason = new FaultReason(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), System.Globalization.CultureInfo.InvariantCulture);
+                secureConversationCloseNotSupportedFault = MessageFault.CreateFault(faultCode, faultReason);
+            }
+            return secureConversationCloseNotSupportedFault;
+        }
+
+        void ThrowIfSecureConversationCloseMessage(Message message)
+        {
+            if (message.Headers.Action == this.secureConversationCloseAction)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new MessageSecurityException(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), null, GetSecureConversationCloseNotSupportedFault()));
+            }
+        }
+       
+        [SecuritySafeCritical]
+        internal SecurityProtocolCorrelationState VerifyIncomingMessage(ref Message message, TimeSpan timeout, params SecurityProtocolCorrelationState[] correlationState)
+        {
+            if (message == null)
+            {
+                return null;
+            }
+            if (this.SecurityProtocol == null)
+                throw new Exception("SecurityProtocol is null");
+            ThrowIfSecureConversationCloseMessage(message);
+           // using (this.ApplyHostingIntegrationContext(message))
+           // {
```

Comment:
You can delete this using. It's used to set the ASP.NET statics such as HttpContext.Current